### PR TITLE
nomad: include snapshot index when submitting plans

### PR DESCRIPTION
**Pull request by schmichael**
# tl;dr

Ensure leader's state snapshot is at or after `max(previousPlanResultIndex, plan.SnapshotIndex)` when evaluating and applying a plan.

# Background

Workers receive evals, create plans, and submit them to the leader for evaluation and applying. This scheduling pipeline runs concurrent with Raft over RPC between server agents to provide optimistic parallelism. The leader is responsible for evaluating workers' plans by ensuring they do not conflict and then applying them. Since this entire process is concurrent with Raft consensus, certain invariants must be met by the leader to ensure plans are evaluated and applied serially:

1. The leader's state used to evaluate plans must be equal to or newer than the state used to create the plan to ensure all referenced objects exist.
2. The leader's state used to evaluate and apply plans must be after previously applied plans (serialization).

# Implementation

The first invariant is enforced by ensuring the state snapshot used to evaluate plans is >= the snapshot at which the plan was created. (new in 1d670f2)

The second invariant is enforced through 3 mechanisms:

* The previous plan is optimistically applied to the leader's state when evaluating a plan. (pre-existing functionality)
* A Raft barrier is emitted after leader election which among other things ensures the leader's state includes all previously applied plans. (pre-existing functionality)
* The leader's state is at or after the previous plan result's index. (new in f82f2a6)

If either invariant cannot be met the plan fails and the worker must reprocess the evaluation.

# Previous Attempt

The previous attempt in #5411 attempted to enforce the invariants by ensuring the leader's state had caught up to the latest Raft index. It also contained some implementation bugs, but the concept was a valid way to enforce the invariants. However, it suffered from a couple issues:

* Certain Raft log entries, such as Barriers, do not increment any indexes in the state store, so the plan applier could never make progress if the previous log entry was a Barrier.
* Waiting for the leader's state to catch up to Raft's LastIndex, even if possible, could cause unnecessary blocking not required to enforce consistency. As long as the state is at a point to provide the invariants above, it may lag behind Raft's LastIndex without affecting correctness.

### CODE REVIEW

File: nomad/plan_apply.go [hunk 1]
```diff
@@ -99,7 +100,16 @@ func (p *planner) planApply() {
 		// Snapshot the state so that we have a consistent view of the world
 		// if no snapshot is available
 		if waitCh == nil || snap == nil {
-			snap, err = p.fsm.State().Snapshot()
+			const timeout = 5 * time.Second
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			snap, err = p.fsm.State().SnapshotAfter(ctx, pending.plan.SnapshotIndex)
```
Comment (notnoop) [inline nomad/plan_apply.go]: We probably should comment why we wait and why it's the `plan.SnapshotIndex` that's worth waiting for.

As a minor nit-pick, I find the name `SnapshotAfter` a bit confusing, it's more `SnapshotWhenAtleast` or `AtOrAfter` rather than `After.

Also, should we handle the case where `snap != nil` but is behind `pending.plan.SnapshotIndex`?  Though, I don't quite follow all of the interactions of loop/waitCh/snap manipulation to fully understand if it's needed.

Comment (schmichael) [reply nomad/plan_apply.go]: > Also, should we handle the case where snap != nil but is behind pending.plan.SnapshotIndex? Though, I don't quite follow all of the interactions of loop/waitCh/snap manipulation to fully understand if it's needed.

Yup! Definitely a bug.

Comment (schmichael) [reply nomad/plan_apply.go]: Renamed SnapshotMinIndex with rationale in commit bcad687 message :

> SnapshotMinIndex was chosen as it describes the action (snapshot), a
constraint (minimum), and the object of the constraint (index).

File: nomad/plan_apply.go [hunk 2]
```diff
@@ -99,7 +100,16 @@ func (p *planner) planApply() {
...
+			cancel()
+			if err == context.DeadlineExceeded {
+				p.logger.Error("timed out synchronizing to planner's index",
+					"timeout", timeout, "plan_index", pending.plan.SnapshotIndex)
+				err = fmt.Errorf("timed out after %s waiting for index=%d",
+					timeout, pending.plan.SnapshotIndex)
```
[the remark this answers is not in the document]
Comment (schmichael) [reply nomad/plan_apply.go]: Rewritten

File: nomad/plan_apply.go [hunk 3]
```diff
@@ -88,18 +100,29 @@ func (p *planner) planApply() {
...
 
-		// Check if out last plan has completed
+		// If last plan has completed get a new snapshot
 		select {
-		case <-waitCh:
-			waitCh = nil
+		case idx := <-planIndexCh:
+			// Previous plan committed. Discard snapshot and ensure
+			// future snapshots include this plan.
+			prevPlanResultIndex = max(prevPlanResultIndex, idx)
```
Comment (notnoop) [inline nomad/plan_apply.go]: Curious if this defensiveness or a possible case?  Can we get a case where the index of the just committed plan is older than that of the previous plan?

Also, how should we handle if `asyncPlanWait` fails and the channel was closed without a value?  I assume the code is fine (as `idx` being 0) , but the comment of plan being committed might not be accurate.

File: nomad/plan_apply.go [hunk 4]
```diff
@@ -88,18 +100,29 @@ func (p *planner) planApply() {
...
+			if idx, err := snap.LatestIndex(); err != nil || idx < minIndex {
+				snap = nil
+			}
+		}
+
 		// Snapshot the state so that we have a consistent view of the world
 		// if no snapshot is available
-		if waitCh == nil || snap == nil {
-			snap, err = p.fsm.State().Snapshot()
+		if planIndexCh == nil || snap == nil {
```
Comment (notnoop) [inline nomad/plan_apply.go]: I feel like I need to build a SAT-solver to understand some of the looping and branching here :P, and given that oder is critical and we cannot substitute them with `else if`, it might be nice explaining the high-level flow in a comment, e.g. why does ch/snap/etc nilness signify in each of the conditionals.

File: nomad/plan_apply.go [hunk 5]
```diff
@@ -141,12 +165,35 @@ func (p *planner) planApply() {
...
+	defer metrics.MeasureSince([]string{"nomad", "plan", "wait_for_index"}, time.Now())
+
+	// Minimum index the snapshot must include is the max of the previous
+	// plan result's and current plan's snapshot index.
+	minIndex := max(prevPlanResultIndex, planSnapshotIndex)
+
+	const timeout = 5 * time.Second
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	snap, err := p.fsm.State().SnapshotMinIndex(ctx, minIndex)
+	cancel()
```
Comment (notnoop) [inline nomad/plan_apply.go]: Should we use a defer for the cancel invocation, it's a little more idomatic and is possible now that you've extracted the function.

File: nomad/plan_apply.go (hunk 3 above)
Comment (schmichael) [reply nomad/plan_apply.go]: Great questions!

> Can we get a case where the index of the just committed plan is older than that of the previous plan?

No.

> Also, how should we handle if asyncPlanWait fails and the channel was closed without a value? I assume the code is fine (as idx being 0) , but the comment of plan being committed might not be accurate.

Yes, the intent of `max(prev, cur)` here is to drop 0 value from failed plan applications. I'll update the comment to reflect that.

**Comment by schmichael** [comment]
> > Plans must also be committed serially, so Plan N+1 should use a state
snapshot containing Plan N. This is guaranteed for plans after the
first plan after a leader election.

> Curious how is that guaranteed now? If two workers submitted independent non-conflicting plans with the same snapshot index N -1; planner would apply first plan and commit with index N, where does it wait until N is commited and applied to State before evaluating the second plan?

If 3 Plans (P1, P2, and P3) are all committed with the same priority and snapshot index the following could occur (note that the order of the plans is arbitrary; concurrent workers could have submitted them in any order).

1. Leader dequeues P1
2. Leader waits until state == P1.SnapshotIndex (100)
3. Leader evaluates and applies P1
4. Leader optimistically adds P1 to state snapshot @ 100
5. Leader dequeues P2, state is already at P2.SnapshotIndex (100)
6. Leader evaluates P2, waits for P1 to commit, waits for state == P1 commit index (101), applies
7. Leader optimistically adds P2 to state snapshot @ 101
8. Leader dequeues P3, state (@101) is already past P3 SnapshotIndex (100)

Leader now evaluates P3 against a state containing both P1 (from index 101) and P2 (from optimistic insert) and the logic proceeds. Even if SnapshotIndex jumps wildly between future and past values (relative to the leader's state index) depending on queue depth, plan priority, and/or worker speed, the optimistic insert and waiting on plan commit index ensures plans are evaluated against a state snapshot containing all previous plans.
